src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -2384,3 +2388,366 @@ case class MatrixFilterEntries(child: MatrixIR, pred: IR) extends MatrixIR {
     mv.copy(rvd = newRVD)
   }
 }
+
+// follows key_by non-empty key
+case class TableAggregateByKey(child: TableIR, expr: IR) extends TableIR {
+  require(child.typ.keyOrEmpty.nonEmpty)
+
+  def children: IndexedSeq[BaseIR] = Array(child, expr)
+  
+  def copy(newChildren: IndexedSeq[BaseIR]): TableAggregateByKey = newChildren match {
+    case Seq(child: TableIR, _: IR) =>
+      TableAggregateByKey(child, expr)
+  }
+  
+  val typ: TableType = child.typ.copy(rowType = child.typ.keyType.get.merge(coerce[TStruct](expr.typ))._1)
+  
+  def execute(hc: HailContext): TableValue = {
+    val prev = child.execute(hc)
+    assert(prev.rvd.isInstanceOf[OrderedRVD])
+    val prevRVD = prev.rvd.asInstanceOf[OrderedRVD]
+
+    val (rvAggs, makeInit, makeSeq, aggResultType, makeAnnotate, rTyp) = ir.CompileWithAggregators[Long, Long, Long, Long](
+      "global", child.typ.globalType,
+      "global", child.typ.globalType,
+      "row", child.typ.rowType,
+      expr,
+      (nAggs, initializeIR) => initializeIR,
+      (nAggs, sequenceIR) => sequenceIR)
+
+    val nAggs = rvAggs.length
+    
+    assert(coerce[TStruct](rTyp) == typ.valueType, s"$rTyp, ${ typ.valueType }")
+
+    val rowType = prev.typ.rowType
+    val keyType = prev.typ.keyType.get
+    val keyIndices = prev.typ.keyFieldIdx.get
+    val keyOrd = prevRVD.typ.kRowOrd
+    val globalsType = prev.typ.globalType
+    val globalsBc = prev.globals.broadcast
+    
+    val newValueType = typ.valueType
+    val newRowType = typ.rowType
+    val newOrvdType = prevRVD.typ.copy(rowType = newRowType)
+    
+    val newRVD = prevRVD.boundary.mapPartitionsPreservesPartitioning(newOrvdType, { (ctx, it) =>
+      val partRVB = new RegionValueBuilder()
+
+      val partRegion = ctx.freshContext.region
+
+      partRVB.set(partRegion)
+      partRVB.start(globalsType)
+      partRVB.addAnnotation(globalsType, globalsBc.value)
+      val partGlobalsOff = partRVB.end()
+
+      val initialize = makeInit()
+      val sequence = makeSeq()
+      val annotate = makeAnnotate()
+
+      new Iterator[RegionValue] {
+        var isEnd = false
+        var current: RegionValue = _
+        val rowKey: WritableRegionValue = WritableRegionValue(keyType, ctx.freshRegion)
+        val consumerRegion: Region = ctx.region
+        val rvb = new RegionValueBuilder()
```

Comment:
I think you can reuse the same region value builder as before. I'd simplify by just having the one RVB and calling it `rvb`.